### ISSUE TITLE
`shell.nix`: how to handle dependencies with binaries?

### ISSUE THREAD

**Issue by jherland** [tweag/FawltyDeps#330]
(I discovered this while attempting to add `ruff` to our project on my NixOS machine.)

### Situation

Our `shell.nix` is intended to quickly create a dev environment from scratch that allows us to run the major Python workflows in our project (Poetry, Nox, etc.) _without_ having to duplicate/translate these into Nix[^1]. As such, it is something of a hybrid or impure setup. When using `nix-shell`:

- Nix provides all supported Python versions along with Poetry itself.
- In the `shellHook` at the end of `shell.nix`, we create a Poetry virtualenv, and
- install all dependencies in the `dev` group. This installation is "impure" in the sense that these dependencies do not come from Nix/Nixpkgs, but are instead retrieved by Poetry (from PyPI). However, since Poetry uses `poetry.lock` (which includes hashes for all dependencies, and is itself version-controlled) there is still ~perfect reproducibility here.
- Finally, we _activate_ the Poetry virtualenv, meaning that the resulting Nix dev shell has all deps at our fingertips, and we can run `python`/`poetry`/`pip`/`pytest`/`nox`/`mypy`/etc. commands directly with no fuzz.

This has worked well until now, even for dependencies that include precompiled code. For example with `mypy`, the installed package include many precompiled `.so` files, and these are indeed used at runtime:
```sh
$ python -c "from mypy import errors; print(errors.__file__)"
/home/jherland/.cache/pypoetry/virtualenvs/fawltydeps-sotR21sh-py3.10/lib/python3.10/site-packages/mypy/errors.cpython-310-x86_64-linux-gnu.so
```
(I believe this works because these `.so` files are dynamlcally loaded into the `python` process, and since `python` itself was provided by Nix, any other shared libs that are needed by these `.so`s will be found in the Nix environment.)

### Problem

**However**, when I tried to add `ruff` to this, things started failing:
```sh
$ pip install ruff  # Temporary, to be replaced by proper install via pyproject.toml
Collecting ruff
  Using cached ruff-0.0.267-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.5 MB)
Installing collected packages: ruff
Successfully installed ruff-0.0.267
$ ruff
bash: /home/jherland/.cache/pypoetry/virtualenvs/fawltydeps-sotR21sh-py3.10/bin/ruff: cannot execute: required file not found
```

The problem is that the `ruff-0.0.267-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl` wheel/package installs a `ruff` binary that is not Nix/NixOS-compatible: As other (non-Nix) Linux executables it encodes an interpreter that is not present on NixOS (and that in any case breaks Nix hermeticity):
```sh
$ patchelf --print-interpreter /home/jherland/.cache/pypoetry/virtualenvs/fawltydeps-sotR21sh-py3.10/bin/ruff
/lib64/ld-linux-x86-64.so.2
```

In other words: The nix-shell allows us to use _non-Nix_ tools to install dependencies, but only as long those dependencies _do not_ contain executable binaries. Since all the necessary executable binaries have been provided _by_ Nix, they all work (and remain hermetic) _inside_ the nix-shell.

### Attempted solution

Since our project dependencies are installed within the `shellHook` of `shell.nix` we could use `patchelf` in the same `shellHook` to fix up any binaries installed by Poetry:

```diff
diff --git a/shell.nix b/shell.nix
index e43cc14..b1243e0 100644
--- a/shell.nix
+++ b/shell.nix
@@ -20,6 +20,11 @@ pkgs.mkShell {
     python39
     python310WithPoetry
     python311
+
+    # Allow installation of binary wheels by (a) providing manylinux2014
+    # support, and (b) patching binaries installed into the Poetry virtualenv.
+    autoPatchelfHook
+    pythonManylinuxPackages.manylinux2014
   ];
   shellHook = ''
     # This is needed to keep python3.7 working while in the same nix-shell
@@ -30,6 +35,8 @@ pkgs.mkShell {
 
     poetry env use "${python310WithPoetry}/bin/python"
     poetry install --sync --with=dev
+    # Patch binaries in the Poetry virtualenv to link against Nix deps
+    autoPatchelf "$(poetry env info --path)"
     source "$(poetry env info --path)/bin/activate"
   '';
 }
```

This works, in that we now get a working `ruff` installation:

```sh
$ patchelf --print-interpreter /home/jherland/.cache/pypoetry/virtualenvs/fawltydeps-sotR21sh-py3.10/bin/ruff
/nix/store/2j8jqmnd9l7plihhf713yf291c9vyqjm-glibc-2.35-224/lib/ld-linux-x86-64.so.2
$ ruff --version
ruff 0.0.265
```

### Problems with attempted solution

This extra `patchelf` step in our `shellHook` does fix up any binaries installed by the previous `poetry install` command. However:
- If the user runs `poetry` to add/upgrade/reinstall dependencies while inside the nix-shell, un-patched binaries are again introduced and will fail to execute until they are patched, either by running `patchELF <venv_dir>` manually, or by adding/updating the `dev` dependency group in `pyproject.toml` and then restarting the entire nix-shell.
- More urgently: When we run `nox` inside the nix-shell, `nox` will install its own virtualenvs (one per session) and then install the appropriate dependencies within. When `ruff` is among these, we - yet again - get an unpatched version of `ruff` installed, but this time, `nox` will (presumably as part of the `lint` session) immediately attempt to _run_ this unpatched `ruff`, and the whole session will fail as a result.

In summary: Solution 1 only "solves" the problem of unpatched binaries as long as they are installed by the `shellHook` in `shell.nix`. Any binaries installed at a later time (e.g. by the user, inside the shell) will fail.

### Other solutions?

(borrowing heavily from https://unix.stackexchange.com/questions/522822/different-methods-to-run-a-non-nixos-executable-on-nixos below.)

I think the approaches to solving this can be roughly grouped in two:

1. Various hacks/approaches to ensure all binaries are patched inside the nix-shell.
   1.1. Push the entire problem onto the user: The nix-shell already provides the `patchELF` shell function to path up binaries under a given directory. It is up to the user to ensure that this is run when needed. This is not particularly friendly, and does not solve the `nox` case above, where there is no user intervention between _installing_ and _running_ these binaries.
   1.2. The Nox case could be solved similarly to what we already do in GitHub CI: For _other_ reasons (caching, speed), we don't want Nox to use its own virtualenvs in CI, but rather (re)use the Poetry virtualenv directly. This is achieved by using the `--no-venv --no-install` flags on the `nox` command line. This would then result in Nox using the Poetry-installed (and _patched_) version of `ruff`.
   1.3. Consider switching the entire shell.nix over to using `poetry2nix` instead. I'm not yet sure how well this works in a `nix-shell` scenario, as opposed to a `nix-build` scenario where `poetry2nix` is used to _build/package_ a project for consumption by Nix. Also, I am not sure if/how this would fix the Nox case.
   1.4. Find some other way to provide/provision the entire dev shell with Nix. This would probably push us away from Nox and maybe even Poetry, at least in the Nix dev shell (this would be instead of, or maintained in parallel with, the developer instructions for non-Nix users).
2. Allowing unpatched binaries to run inside the nix-shell. This -- to some extent -- accepts the impurity/non-hermiticity introduced by running the unpatched binaries.
   2.1. On non-NixOS (in other words when `/lib64/ld-linux-x86-64.so.2` already exists) I suspect the unpatched binary might run without any problems at all (especially if it's a static executable -- which `ruff` appears to be).
   2.2. On NixOS, there is [nix-ld](https://github.com/Mic92/nix-ld), which should work as long as we setup `$NIX_LD` (and `$NIX_LD_LIBRARY_PATH`?) beforehand. The environment variables could (should?) be set up in our `shell.nix`, but the installation of nix-ld itself is a system-wide thing that must be done by the NixOS user prior entering the nix-shell.
   2.3. Alternatives to nix-ld include [nix-alien](https://github.com/thiagokokada/nix-alien) or steam-run. I have no experience with either of these.
   2.4. Use `buildFHSUserEnv` to create a more encompassing dev environment where unpatched binaries can run on a Nix-provided FHS-like system. I also have little experience here, and am wary of potential problems caused by the mechanism itself.

There may be more alternative solutions in either categories, or even more categories altogether, but this is where I'm at for now. Brain dump complete.

[^1]: The ultimate intention is to have a Python project that follows Python best practices and provides the nix-shell as a value-add _without_ forcing contributors into using Nix.

**Comment by jherland** [comment]
I came across https://jonathanlorimer.dev/posts/nix-thesis.html today, and started thinking about this issue while reading it.

As covered in the [Nix Principles](https://jonathanlorimer.dev/posts/nix-thesis.html#nix-principles) section: "User environments are not a composition mechanism". Nox/`pip install` appears to break this principle when it installs new binaries at runtime, binaries that make assumptions . As mentioned in the [next section](https://jonathanlorimer.dev/posts/nix-thesis.html#topics-not-covered) Nix has a potential workaround that applies here as well: Patching binaries directly, in our case, using `patchElf` to rewrite the dynamic loader.

The problem with us using this approach is that Nox does not offer a hook to patch the `pip install`ed binaries before they are called from within the Nox session.

We could solve this in FawltyDeps by extending our [Nox installation helper](https://github.com/tweag/FawltyDeps/blob/e5843daf90a86c74a028ce2a0f8e321e2acf463e/noxfile.py#L10) (that today exists to handle Poetry dependency groups) to also provide a hook point _after_ the installing is done (but _before_ the rest of the Nox session is run). The hook will run `patchElf` on the installed packages (or across the entire virtualenv) to fix the dynamic loader in any binaries that happened to be installed (e.g. `ruff`).

Consider this alternative 1.5 in the above list.

Remaining questions at this point:
- Do we also need to fixup shared lib references in the installed binaries? In general, how does this workaround interoperate with the `manylinux` definition?
- Are there _other_ cases where we `pip install` and might have to patch up more binaries?
- More generally, does this mean that _any_ language-specific (or other) package installer is fundamentally incompatible with Nix/NixOS?
  - Is there _any_ potential for solutions that allows a user/developer to keep using `pip install` in a way that does not violate Nix/NixOS principles?
  - How "pure"/tractable is it to provide `manylinux` installed into a `FHSUserEnv` and then have `pip install`ed binaries run within this env hopefully(?) without this environment affecting the other programs run within our nix-shell?

**Comment by jherland** [comment]
I'm closing this as solved by PR #422.
